controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -294,13 +294,13 @@ func extractImageVersion(tag string) (string, error) {
 // Some of the checks come from
 // https://github.com/coredns/corefile-migration/blob/v1.0.6/migration/migrate.go#L414
 func validateCoreDNSImageTag(fromVersion, toVersion string) error {
-	from, err := semver.Parse(fromVersion)
+	from, err := util.ParseMajorMinorPatch(fromVersion)
 	if err != nil {
-		return errors.Wrapf(err, "failed to semver parse %q", fromVersion)
+		return errors.Wrapf(err, "failed to parse CoreDNS current version %q", fromVersion)
 	}
-	to, err := semver.Parse(toVersion)
+	to, err := util.ParseMajorMinorPatch(toVersion)
 	if err != nil {
-		return errors.Wrapf(err, "failed to semver parse %q", toVersion)
+		return errors.Wrapf(err, "failed to parse CoreDNS target version %q", toVersion)
 	}
 	if from.Compare(to) >= 0 {
```

Comment:
LGTM
but note that we recently discussed this on a kubeadm PR and we are facing a similar problem:
https://github.com/kubernetes/kubernetes/pull/88845

comparing MAJOR.MINOR.PATCH ignores vendor revisions:
`v1.17.4-somevendor.0` will be considered equal to `v1.17.4-somevendor.1`

we need to standardize these, but it will be tricky:
https://github.com/kubernetes/kubeadm/issues/2058